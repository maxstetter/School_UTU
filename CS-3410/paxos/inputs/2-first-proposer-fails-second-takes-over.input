initialize 3 nodes

// node 3 starts out not receiving any message or responding to anything

// node 1 starts a proposal
at 1000 send prepare request from 1

// deliver two of the three prepare requests, each triggering a response
at 1010 deliver prepare request message to 1 from time 1000
at 1011 deliver prepare request message to 2 from time 1000

// deliver both prepare responses, triggering an accept round
at 1020 deliver prepare response message to 1 from time 1010
at 1021 deliver prepare response message to 1 from time 1011

// accept requests are received
at 1030 deliver accept request message to 1 from time 1021
at 1031 deliver accept request message to 2 from time 1021

// oh no! node 1 froze up and all messages to it were lost
// consensus was achieved (nodes 1 and 2 both voted for it),
// but was not discovered

// node 3 wakes up and decides to send out a proposal
at 1040 send prepare request from 3

// nodes 2 and 3 get the prepare requests
at 1050 deliver prepare request message to 2 from time 1040
at 1051 deliver prepare request message to 3 from time 1040

// the prepare responses are delivered and the propose round succeeds
// (node 3's sequence number beats node 1's sequence number)
at 1060 deliver prepare response message to 3 from time 1050
at 1061 deliver prepare response message to 3 from time 1051

// the accept round starts, but node 3 had to pick node 1's value
// after discovering it in the prepare response from node 2
// the consensus value is preserved because of overlapping quorums
at 1070 deliver accept request message to 2 from time 1061
at 1071 deliver accept request message to 3 from time 1061

// nodes 2 and 3 vote to accept and consensus is achieved again,
// with the same value as before
at 1080 deliver accept response message to 3 from time 1070
at 1081 deliver accept response message to 3 from time 1071

// node 3 learns about the consensus value and notifies everyone
at 1090 deliver decide request message to 2 from time 1081
at 1091 deliver decide request message to 3 from time 1081

// node 1 restarts and tries again
at 1100 send prepare request from 1

// deliver all three prepare requests, each triggering a response
at 1110 deliver prepare request message to 1 from time 1100
at 1111 deliver prepare request message to 2 from time 1100
at 1112 deliver prepare request message to 3 from time 1100

// all three responses include value 11111 since all have voted for
// it in a previous round
at 1120 deliver prepare response message to 1 from time 1110
at 1121 deliver prepare response message to 1 from time 1111
at 1122 deliver prepare response message to 1 from time 1112

// node 1 sends out accept requests (after 2nd vote received)
// using discovered value 11111 with sequence 5003 (the most recent vote)
at 1130 deliver accept request message to 1 from time 1121
at 1131 deliver accept request message to 2 from time 1121
at 1132 deliver accept request message to 3 from time 1121

// all three nodes accept (third vote will be ignored)
at 1140 deliver accept response message to 1 from time 1130
at 1141 deliver accept response message to 1 from time 1131
at 1142 deliver accept response message to 1 from time 1132

// decide messages sent at time 1141 and delivered without incident
at 1150 deliver decide request message to 1 from time 1141
at 1151 deliver decide request message to 2 from time 1141
at 1152 deliver decide request message to 3 from time 1141

// node 1 learns about the decided value for the first time, nodes
// 2 and 3 have it confirmed
