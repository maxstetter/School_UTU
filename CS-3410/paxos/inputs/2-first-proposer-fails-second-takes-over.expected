--> initialized 3 nodes
--> sent prepare requests to all nodes from 1 with sequence 5001
--> prepare request from 1 sequence 5001 accepted by 1 with no value
--> prepare request from 1 sequence 5001 accepted by 2 with no value
--> positive prepare response from 1 sequence 5001 recorded by 1 with no value
--> positive prepare response from 2 sequence 5001 recorded by 1 with no value
--> prepare round successful: 1 proposing its own value 11111
--> sent accept requests to all nodes from 1 with value 11111 sequence 5001
--> accept request from 1 with value 11111 sequence 5001 accepted by 1
--> accept request from 1 with value 11111 sequence 5001 accepted by 2
--> note: consensus has been achieved
--> sent prepare requests to all nodes from 3 with sequence 5003
--> prepare request from 3 sequence 5003 accepted by 2 with value 11111 sequence 5001
--> prepare request from 3 sequence 5003 accepted by 3 with no value
--> positive prepare response from 2 sequence 5003 recorded by 3 with value 11111 sequence 5001
--> positive prepare response from 3 sequence 5003 recorded by 3 with no value
--> prepare round successful: 3 proposing discovered value 11111 sequence 5001
--> sent accept requests to all nodes from 3 with value 11111 sequence 5003
--> accept request from 3 with value 11111 sequence 5003 accepted by 2
--> accept request from 3 with value 11111 sequence 5003 accepted by 3
--> note: consensus has been achieved
--> positive accept response from 2 sequence 5003 recorded by 3
--> positive accept response from 3 sequence 5003 recorded by 3
--> accept round successful: 3 detected consensus with value 11111
--> sent decide requests to all nodes from 3 with value 11111
--> recording consensus value 11111 at 2
--> recording consensus value 11111 at 3
--> sent prepare requests to all nodes from 1 with sequence 5011
--> prepare request from 1 sequence 5011 accepted by 1 with value 11111 sequence 5001
--> prepare request from 1 sequence 5011 accepted by 2 with value 11111 sequence 5003
--> prepare request from 1 sequence 5011 accepted by 3 with value 11111 sequence 5003
--> positive prepare response from 1 sequence 5011 recorded by 1 with value 11111 sequence 5001
--> positive prepare response from 2 sequence 5011 recorded by 1 with value 11111 sequence 5003
--> prepare round successful: 1 proposing discovered value 11111 sequence 5003
--> sent accept requests to all nodes from 1 with value 11111 sequence 5011
--> positive prepare response from 3 sequence 5011 recorded by 1 with value 11111 sequence 5003
--> valid prepare vote ignored by 1 because round is already resolved
--> accept request from 1 with value 11111 sequence 5011 accepted by 1
--> accept request from 1 with value 11111 sequence 5011 accepted by 2
--> note: consensus has been achieved
--> accept request from 1 with value 11111 sequence 5011 accepted by 3
--> positive accept response from 1 sequence 5011 recorded by 1
--> positive accept response from 2 sequence 5011 recorded by 1
--> accept round successful: 1 detected consensus with value 11111
--> sent decide requests to all nodes from 1 with value 11111
--> positive accept response from 3 sequence 5011 recorded by 1
--> valid accept vote ignored by 1 because round is already resolved
--> recording consensus value 11111 at 1
--> recording consensus value 11111 at 2
--> recording consensus value 11111 at 3
